### PR TITLE
AWS API Gateway: Fix API Gateway naming convention

### PR DESCRIPTION
**Pull request by fabsrc**
The naming of API Gateway will be changed from `${stage}-${service}` to `${service}-${stage}` in the next major version.
    
This commit adds a deprecation notice and an option to already adapt to this this convetion.

Adresses: #2918

### CODE REVIEW

**Review by medikoo — CHANGES_REQUESTED** [review]
Thanks @fabsrc for giving that a spin!

Sorry for not having that clear in origin issue, but we cannot introduce breaking changes just like that. All breaking changes are first introduced as deprecation, and then when new major is about to be released we issue a PR's that make changes actually breaking.

I've updated the main description of https://github.com/serverless/serverless/issues/2918 with implementation proposal. Please see notes over there, on how it should be approached.

Thank you!

**Comment by fabsrc** [comment]
Hey @medikoo.

Thanks for the detailed description how to implement this. Makes a lot of sense to do it that way.
I added the deprecation notice and an option to enable the new convention already.

Could you please have another look?

File: lib/plugins/aws/lib/naming.js [hunk 1]
```diff
@@ -252,6 +252,19 @@ module.exports = {
...
+    if (
+      this.provider.serverless.service.provider.apiGateway &&
+      this.provider.serverless.service.provider.apiGateway.shouldStartNameWithService === true
+    ) {
+      return `${this.provider.serverless.service.service}-${this.provider.getStage()}`;
+    }
+
+    this.provider.serverless._logDeprecation(
+      'AWS_API_GATEWAY_NAME_STARTING_WITH_SERVICE',
+      'Starting with next major version, API Gateway naming will be changed from `${service}-${stage}` to `${stage}-${service}`.'
```
Comment (medikoo) [inline lib/plugins/aws/lib/naming.js:265]: Let's rephrase it to:

```
Starting with next major version, API Gateway naming will be changed from "{stage}-{service}" to "{service}-{stage}"
Set "provider.apiGateway.shouldStartNameWithService" to "true" to  adapt to new behavior now
```

File: lib/plugins/aws/provider/awsProvider.js [hunk 2]
```diff
@@ -428,7 +428,13 @@ class AwsProvider {
         },
         provider: {
           properties: {
-            apiGateway: { type: 'object', properties: { websocketApiId: { type: 'string' } } },
+            apiGateway: {
+              type: 'object',
+              properties: {
+                websocketApiId: { type: 'string' },
+                shouldStartNameWithService: { type: 'boolean' },
```
Comment (medikoo) [inline lib/plugins/aws/provider/awsProvider.js:435]: Let's set it as `const: true`. If we would allow also `false`, it'll mean that we should keep support for old way of things in v3.

Comment (medikoo) [inline lib/plugins/aws/provider/awsProvider.js:435]: Let's define properties in alphabetical order

**Review by medikoo — CHANGES_REQUESTED** [review]
Thank you @fabsrc ! That looks really good. I've proposed few final improvements

File: docs/providers/aws/guide/serverless.yml.md [hunk 3]
```diff
@@ -83,6 +83,7 @@ provider:
     binaryMediaTypes: # Optional binary media types the API might return
       - '*/*'
     metrics:  false # Optional detailed Cloud Watch Metrics
+    shouldStartNameWithService: true # Use `${service}-${stage}` naming for API Gateway. Will be `true` by default in next major version.
```
Comment (medikoo) [inline docs/providers/aws/guide/serverless.yml.md:86]: In this doc, we need to expose defaults. So let's keep `false` (even though we explictly do not suport such setting). 
Technically we could set `null`, as that'll be valid on schema grounds, but then again it'll be confusing

File: lib/plugins/aws/lib/naming.js [hunk 4]
```diff
@@ -252,6 +252,20 @@ module.exports = {
     ) {
       return `${this.provider.serverless.service.provider.apiName}`;
     }
+
+    if (
+      this.provider.serverless.service.provider.apiGateway &&
+      this.provider.serverless.service.provider.apiGateway.shouldStartNameWithService === true
```
Comment (medikoo) [inline lib/plugins/aws/lib/naming.js:258]: Let's write it as:

```javascript
if (_.get(this.provider.serverless.service.provider.apiGateway, "shouldStartNameWithService")) {
..
```

It's not natural in JS to check for `true` explicitly

**Review by medikoo — COMMENTED** [review]
Thank you @fabsrc, looks very good. Just few last style suggestions

**Review by medikoo — COMMENTED** [review]
@fabsrc thanks for update. Can you rebase with master? Currently PR has some conflicts and is not in mergeable state

**Review by medikoo — APPROVED** [review]
Thank you @fabsrc !

**Comment by Poonaka** [comment]
Will there be a way to keep the old naming convention in future releases?

**Comment by medikoo** [comment]
> Will there be a way to keep the old naming convention in future releases?

Thanks for asking. Currently we want to drop support for old naming intermittently with new version, still we're open for feedback on that. 

Do you find it problematic if v3 will not support old naming?

**Comment by Poonaka** [comment]
> > Will there be a way to keep the old naming convention in future releases?
> 
> Thanks for asking. Currently we want to drop support for old naming intermittently with new version, still we're open for feedback on that.
> 
> Do you find it problematic if v3 will not support old naming?

We used serverless to setup some Cognito user pools and lambda triggers that occasionally need to be updated. Just wondering how we will migrate to the new naming convention without serverless deleting the user pools.

**Comment by fabsrc** [comment]
> Will there be a way to keep the old naming convention in future releases?

I think by explicitly setting the name of the API Gateway in the provider settings (`provider.apiName`) you can keep the old convention. E.g.

```yaml
provider:
  name: aws
  apiName: ${opt:stage, 'dev'}-${self:service.name}
```

**Comment by Poonaka** [comment]
> > Will there be a way to keep the old naming convention in future releases?
> 
> I think by explicitly setting the name of the API Gateway in the provider settings (`provider.apiName`) you can keep the old convention. E.g.
> 
> ```yaml
> provider:
>   name: aws
>   apiName: ${opt:stage, 'dev'}-${self:service.name}
> ```

Perfect, we will give that a shot. Thanks!

**Comment by medikoo** [comment]
@fabsrc that's a great hint. Also using `apiName` will turn off the deprecation notice. So that technically the recommended way to carry on without change in next major. I'll add that info to deprecation notes
